rfcs/InputUnion.md

```diff
@@ -133,11 +133,31 @@ In this mutation, we encounter the main challenge of the **Input Union** - we ne
 
 A wide variety of solutions have been explored by the community, and they are outlined in detail in this document under [Possible Solutions](#Possible-Solutions).
 
-## Requirements
+## Solution Shape - Aims & Objections
 
-1. **Doesn't inhibit [schema evolution](https://graphql.github.io/graphql-spec/draft/#sec-Validation.Type-system-evolution)** ─ adding a new member type to an Input Union or doing any non-breaking change to existing member types must not result in breaking change. For example, adding a new optional field to member type or changing a field from non-nullable to nullable must not break previously valid client operations.
+Hypothetical goals that a solution might attempt to fulfill.
 
-**Note**: Input Unions may enforce some restrictions on member types, but if they do so compliance with these restrictions must be fully validated during schema building (analagous to how interfaces enforce restrictions on member types).
+### GraphQL should contain an Input Union type
+
+The premise of this RFC - GraphQL should contain a polymorphic Input Union type.
+
+### Doesn't inhibit [schema evolution](https://graphql.github.io/graphql-spec/draft/#sec-Validation.Type-system-evolution)
+
+Adding a new member type to an Input Union or doing any non-breaking change to existing member types does not result in breaking change. For example, adding a new optional field to member type or changing a field from non-nullable to nullable does not break previously valid client operations.
+
+### Any member type restrictions are validated in schema
+
+If a solution places any restrictions on member types, compliance with these restrictions should be fully validated during schema building (analagous to how interfaces enforce restrictions on member types).
+
+### A member type may be a Scalar
+
+In addition to containing Input types, an Input Union may also contain Scalars.
+
+### A member type may be an Enum
+
+In addition to containing Input types, an Input Union may also contain `Scalar`s or `Enum`s.
+
+* Objection: An input union containing both a `String` and an `Enum` would be impossible to distinguish.
```

Comment:
I would merge `Scalar` and `Enum` into `Leaf type` especially since we use this term through GraphQL spec so it's not something new.
As for objection it also applied to scalar types since `ID` can intersect with `String`/`Int` and `Float` can intersect with `Int`. 
Also, I don't think it's objection since it just requires boxing leaf values e.g. `@oneOf` solves this issues `{ string: 'foo' }`, `{ enum: foo }`.

Reply:
The spec isn't explicit on this, but the term "leaf" may also be interpretted include lists of scalars and enums since these also do not have selection sets. I think we want to explicitly preclude lists from this particular title. If lists in the input union (as opposed to lists OF the input union) are desired I'd suggest adding another header for that.

Reply:
I agree with @IvanGoncharov, in the space of all solutions, there are definitely some that allow the inclusion of leaf types.

Reply:
A stronger objection IMO (or at least a different objection) is that output polymorphism (Union and Interface types) don't support leaf types/scalars/etc, and if we want to support that in polymorphism generally we should do it for both.

Reply:
(@eapache I agree, and `@oneField` can be extended to do this very easily. I've left it out for now to keep the proposal small and focussed.)

Reply:
merged those two & added that additional objection